source/PluginHelper.cpp

```diff
@@ -0,0 +1,198 @@
+/* PluginHelper.cpp
+Copyright (c) 2023 by RisingLeaf(https://github.com/RisingLeaf)
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "PluginHelper.h"
+
+#if defined _WIN32
+#define WIN32_LEAN_AND_MEAN
+#endif
+
+#include <archive.h>
+#include <archive_entry.h>
+#include <cstring>
+#include <curl/curl.h>
+#include <sys/stat.h>
+#include <stdio.h>
+
+#if defined _WIN32
+#include "text/Utf8.h"
+#include <windows.h>
+#else
+#include <unistd.h>
+#endif
+
+using namespace std;
+
+
+
+namespace PluginHelper {
+	// Max size for extracting an archive in bytes this will be 1GB.
+	const int MAX_SIZE = 1000000000;
+
+	size_t WriteData(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+		size_t written = fwrite(ptr, size, nmemb, stream);
+		return written;
+	}
+
+
+
+	bool Download(string url, string location)
+	{
+		CURL *curl = curl_easy_init();
+		if(!curl)
+			return false;
+#if defined _WIN32
+		FILE *out = nullptr;
+		_wfopen_s(&out, Utf8::ToUTF16(location).c_str(), L"wb");
+#else
+		FILE *out = fopen(location.c_str(), "wb");
+#endif
+		// Set the url that gets downloaded
+		curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+		// Follow redirects
+		curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1l);
+		// How long we will wait
+		curl_easy_setopt(curl, CURLOPT_CA_CACHE_TIMEOUT, 604800L);
+		// What is the maximum filesize in bytes.
+		curl_easy_setopt(curl, CURLOPT_MAXFILESIZE, MAX_SIZE);
+		// Set the write function and the output file used in the write function
+		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteData);
```

Comment:
It might be possible to pass `fwrite` directly instead. I'm not sure though

Reply:
that worked